compiler/tests/prepositionGenerator.bal

```diff
@@ -0,0 +1,457 @@
+import ballerina/io;
+import ballerina/test;
+import ballerina/time;
+
+import wso2/nballerina.types as t;
+
+// diff-2 stand for isEmpty(diff(a, b)) and isEmpty(diff(b, a))
+enum PropositionOp {
+    UNION, INTERSECT, DIFF, DIFF2
+}
+
+type Proposition record {|
+    PropositionOp op;
+    boolean isEmpty;
+    int left;
+    int right;
+|};
+
+type SubtypeProposition record {|
+    *Proposition;
+    DIFF op = DIFF;
+    true isEmpty = true;
+|};
+
+type NonEmptyProposition record {|
+    *Proposition;
+    UNION op = UNION;
+    false isEmpty = false;
+|};
+
+type PropositionGenerator function (Context cx, PropositionPath path) returns Proposition;
+
+type SubtypePropositionGenerator function (Context cx, PropositionPath path) returns SubtypeProposition;
+
+type NonEmptyPropositionGenerator function (Context cx, PropositionPath path) returns NonEmptyProposition; 
+
+type Limits readonly & record {|
+    readonly int maxTupleMembers = 6;
+    readonly int maxStringConstLen = 20;
+    readonly int duplicationFactor = 6;
+|};
+
+type PropositionListCtor function () returns Proposition[];
+
+type OnFail function(Context cx, Proposition failedProp);
+
+class Context {
```

Comment:
These names are module-scoped, so I think we need a more specific name, such as GeneratorContext. Some other names here have the same problem.